src/Symfony/Bundle/SecurityBundle/Security/FirewallConfig.php

```diff
@@ -0,0 +1,126 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\SecurityBundle\Security;
+
+/**
+ * @author Robin Chalas <robin.chalas@gmail.com>
+ */
+class FirewallConfig
+{
+    private $name;
+    private $requestMatcher;
+    private $securityEnabled;
+    private $stateless;
+    private $provider;
+    private $context;
+    private $entryPoint;
+    private $accessDeniedHandler;
+    private $accessDeniedUrl;
+    private $userChecker;
+    private $listeners;
+
+    public function __construct($name, $requestMatcher, $securityEnabled = true, $stateless = false, $provider = null, $context = null, $entryPoint = null, $accessDeniedHandler = null, $accessDeniedUrl = null, $userChecker = null, $listeners = array())
```

Comment:
I started with exactly what you propose and came to this. 
Having distinct arguments is imho more structuring and gives more visibility about what the class provides. Also it avoids lots of `isset($key) ? $key : null`, we would have to build `$config` from `SecurityExtension` anyway.

Also, please see #19398 for this part, this depends on :)